Refactor carousel to use swiper.js

glendaviesnz opened the pull request:
Fixes #15678 
Fixes #6069 
Fixes #18923 
Fixes #9011 


Currently the touch screen usability of the carousel module is not as good as other online galleries, or the Jetpack slideshow block.

This PR removes the custom carousel functionality and touch screen utils and replaces them with the 3rd party [swiper.js](https://swiperjs.com/) library which is used by the Jetpack slideshow block.

![2021-06-24 20 34 15](https://user-images.githubusercontent.com/1464705/123365698-a2688180-d52b-11eb-831c-1dd1f9ab9adc.gif)


#### Testing instructions:
1. Checkout this PR in local jetpack dev env
2. Enable the Jetpack carousel option in jetpack settings
2. Add a gallery to a page or post, load in front end and check that the carousel lightbox loads and works as expected
3. On a mobile device run the gallery and check that swipe works well, and also pinch zoom and pan

#### Ensure existing Carousel features and toggles work as expected
1. Go to `/wp-admin/admin.php?page=jetpack#/writing` and toggle showing exif and comments (each setting should show/hide icons in the bottom right of the carousel)
2. From the settings above, switch to the White carousel color scheme and test that it works correctly

#### Does this pull request change what data or activity we track or use?
No

ramonjd commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:
```diff
@@ -1628,8 +1171,31 @@
-			selectSlideAtIndex( settings.startIndex );
+			carousel.overlay.style.display = 'block';
+
+			swiper = new Swiper( '.swiper-container', {
```
I guess we could pass in `window.Swiper` to the self-executing anon func to keep the linter quiet.

glendaviesnz replied:
yeh, or Andrew's work on making this es6 modules might be a better way around this - but that could be a second phase.

andrewserong replied:
> Andrew's work on making this es6 modules might be a better way around this

I think the es6 modules approach seems promising (and I'm [obviously pretty excited](https://github.com/Automattic/jetpack/pull/20131) about it), but agreed we could totally tackle that in a follow-up if it looks like swiper.js isn't too heavy to pull in directly (especially since we're already doing that for the Slideshow block?).

ramonjd commented:
Tested this on an iOs device (Safari and Chrome) and simulator (to record the screen).

![Jun-21-2021 13-36-06](https://user-images.githubusercontent.com/6458278/122704060-3ae5d400-d296-11eb-9a70-f2feddfd4ed6.gif)


It's already a thousand times better when swiping through the images. ❤️ 

Pinch and zoom works for me, with a double-tap returning to the original zoom.

You've probably got this on your radar, but one thing we might like to do is sit the navigation dots closer to the image or resize the gallery relative to the viewport dimensions. It's hidden on on iPhone 11.


![IMG_FCD2AE347F0E-1](https://user-images.githubusercontent.com/6458278/122704374-ec850500-d296-11eb-9150-8facd1529df0.jpeg)

glendaviesnz commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:
```diff
@@ -904,95 +633,19 @@
-				setSlidePosition( next.el, Math.ceil( galleryWidth - screenPadding * 0.75 ) );
-				domUtil.show( next.el );
-			}
+			// Clear slide data for DOM garbage collection.
```
should be able to use the old fadeOut method here, but couldn't get it working for some reason, so just hacked in a manual close of the overall for now - need to come back to this.

glendaviesnz commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:
```diff
@@ -1417,35 +996,6 @@
 			}
 		}
 
-		function loadFullImage( slide ) {
```
need to add this back in so we are getting the best quality image for current slide.

glendaviesnz replied:
done

glendaviesnz commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:
```diff
@@ -1628,8 +1167,31 @@
+			swiper.on( 'slideChange', function () {
+				selectSlideAtIndex( swiper.activeIndex );
+			} );
+			carousel.overlay.style.opacity = '1';
```
should be able to use existing fadeIn method for this but wasn't working for some reason, so temporarily adjusting here and will come back to this.

sgomes replied:
Let me know if you need a hand with that when you look into it later. I wrote that method to replace `jQuery` functionality in a more performant way, but I could have missed something.

glendaviesnz replied:
This was my mistake - had set a `display:none` in css in stead as inline style - fixed now.

glendaviesnz commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.php:
```diff
@@ -239,6 +239,14 @@ function enqueue_assets() {
 				true
 			);
 
+			wp_enqueue_script(
```
once we finalise if we are going ahead with this will refactor this to a local import.

apeatling commented:
This looks like a significant improvement @glendaviesnz, let's keep pushing on this. Doesn't look like the zoom and pan support is turned on here? I'm just getting the basic browser zooming.

glendaviesnz commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:
```diff
@@ -1274,7 +1274,18 @@
+					var jsScript = document.createElement( 'script' );
+					jsScript.id = 'jetpack-carousel-js';
+					jsScript.type = 'text/javascript';
+					jsScript.src = '/wp-content/plugins/jetpack/modules/carousel/swiper-bundle.js';
```
need to work out how to generate this path properly based on env

andrewserong replied:
You could always pass in the path to it from the PHP side? We already add the login URL to `login_url` in `$localize_strings` so there's a precedent there 🙂

glendaviesnz replied:
managed to do it with `wp_localize_script`

apeatling reviewed (approved):
Tested this on desktop and mobile and it loads and works correctly for me. There are obviously design improvements that need to be made, but I'd like to get this in so we can follow up quickly with more PRs to do that. 

@jeherve what do you think about pushing this through so we can get it all polished up in followup PRs before June 28?

jeherve commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.php:
```diff
@@ -320,6 +328,15 @@ function enqueue_assets() {
+			wp_enqueue_style(
+				'jetpack-carousel-swiper-css',
+				Assets::get_file_url_for_environment(
+					'_inc/build/carousel/swiper-bundle.min.css',
```
For the bundle to be minified, I believe we'll need to add its reference here:
https://github.com/Automattic/jetpack/blob/e13487d0dbbe4c743739b6f1a3e37c143a8c2614/projects/plugins/jetpack/tools/builder/frontend-css.js#L25
https://github.com/Automattic/jetpack/blob/b1c8b8ecba896ab616a2d7ebe43ecf5d6a6463db/projects/plugins/jetpack/class.jetpack.php#L63

glendaviesnz replied:
Perfect, working out how to get the css into the build process was on my todo list todo, that saves me some hunting! Done.

jeherve commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.php:
```diff
@@ -239,6 +239,14 @@ function enqueue_assets() {
 				true
 			);
 
+			$swiper_library_path = array(
+				'url' => Assets::get_file_url_for_environment(
+					'_inc/build/carousel/swiper-bundle.min.js',
+					'modules/carousel/swiper-bundle.js'
+				),
+			);
+			wp_localize_script( 'jetpack-carousel', 'jetpackSwiperLibraryPath', $swiper_library_path );
```
Since we do not enqueue the file directly, will this work with plugins that concatenate JS, like Autoptimize (enabled on some WoA sites)?

glendaviesnz replied:
I may be misunderstanding what you are saying here, but the `wp_localize_script` is running against the existing the existing `jetpack-carousel` script which is enqueued?

glendaviesnz replied:
FYI, I ran Autoptimize locally with the js and css options toggled on and everything still worked as expected.

jeherve reviewed (commented):
> what do you think about pushing this through so we can get it all polished up in followup PRs before June 28? We have a number in progress that are blocked on this.

I don't see why not. I would, however, be happy to have a review from @sgomes since he worked on performance improvements for the Carousel feature in the past few months, and will most likely have useful feedback.

sgomes commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:
```diff
@@ -1546,41 +1117,58 @@
 					slideEl.setAttribute( 'itemscope', '' );
 					slideEl.setAttribute( 'itemtype', 'https://schema.org/ImageObject' );
-					domUtil.hide( slideEl );
+					// domUtil.hide( slideEl );
```
See above.

glendaviesnz replied:
deleted.

sgomes commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:
```diff
@@ -1546,41 +1117,58 @@
+					zoomEl.classList.add( 'swiper-zoom-container' );
 
-					slideEl.style.left = i < startIndex ? -1000 : carousel.gallery.offsetWidth;
+					// slideEl.style.left = i < startIndex ? -1000 : carousel.gallery.offsetWidth;
```
See above.

glendaviesnz replied:
deleted

sgomes commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:
```diff
@@ -1546,41 +1117,58 @@
+			if ( ! window.Swiper ) {
+				var loader = document.querySelector( '#jp-carousel-loading-overlay' );
+				domUtil.show( loader );
+				var head = document.getElementsByTagName( 'head' )[ 0 ];
```
`document.head` has been supported since IE 9, no need for querying here.

glendaviesnz replied:
Thanks, have simplified this now.

sgomes commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:
```diff
@@ -1546,41 +1117,58 @@
+				var head = document.getElementsByTagName( 'head' )[ 0 ];
+				var jsScript = document.createElement( 'script' );
+				jsScript.id = 'jetpack-carousel-swiper-js';
+				jsScript.type = 'text/javascript';
```
No need for this, every browser assumes JS.

glendaviesnz replied:
Removed.

sgomes commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:
```diff
@@ -1546,41 +1117,58 @@
+				domUtil.show( loader );
+				var head = document.getElementsByTagName( 'head' )[ 0 ];
+				var jsScript = document.createElement( 'script' );
+				jsScript.id = 'jetpack-carousel-swiper-js';
```
This is fine, but I just wanted to point out that it's not an issue to have the script tag without an ID 👍

glendaviesnz replied:
I was tossing up whether to checking for the existing of this element via id before adding to avoid appending more than once, but looking at the code path I think the check for Swiper object should prevent this from happening - but will do some more testing today with double clicking, etc. to make sure there isn't any race condition that could cause a double appending of the script, if not I will remove the id.

sgomes commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:
```diff
@@ -1546,41 +1117,58 @@
+				var jsScript = document.createElement( 'script' );
+				jsScript.id = 'jetpack-carousel-swiper-js';
+				jsScript.type = 'text/javascript';
+				jsScript.src = window.jetpackSwiperLibraryPath.url;
```
We should add `jsScript.async = true;` so that the external script gets loaded asynchronously and doesn't potentially block the main thread.

glendaviesnz replied:
done

sgomes reviewed (dismissed):
Thank you for working on this, @glendaviesnz!

Here's a first review pass by only looking at the code. I'll be running some testing later today, and will add other comments if I find any issues 👍

glendaviesnz commented on projects/plugins/jetpack/changelog/refactor-carousel-to-use-swiper:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: major
+
+Replaces the custom carousel js code with the swiper js library
```
Good point, have updated.

glendaviesnz commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:
```diff
@@ -1546,41 +1117,58 @@
-					image.style.width = '100%';
-					image.style.height = '100%';
+					image.src = attrs.src;
+					// image.style.width = '100%';
```
Woops, thanks, missed all of these when tidying up my work! Gone now.

ramonjd commented:
I've just tested on the following devices:


- Iphone 8 Safari
- Iphone 12 Safari
- Samsung 21 Chrome
- Google Pixel Chrome 
- Desktop Edge
- Desktop Safari
- Desktop Firefox
- Desktop Chrome

All work as expected 🎉 , though on the mobile devices the dots navigation is mostly hidden. Hoping to resolve those sorts of issues in https://github.com/Automattic/jetpack/pull/20139

ramonjd commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:
```diff
@@ -1448,6 +1012,21 @@
+				currentSlide.style.backgroundImage = 'url(' + slide.attrs.mediumFile + ')';
+				return;
+			}
+			slide.el.style.backgroundImage = 'url(' + slide.attrs.mediumFile + ')';
```
Assuming this is a pre-loader?

Maybe we can add some styles to the active element to hide it?

<img width="300" alt="Screen Shot 2021-06-23 at 3 58 48 pm" src="https://user-images.githubusercontent.com/6458278/123043628-eda36700-d43b-11eb-9b8b-852e180cae26.png">

E.g., 

```css
.swiper-slide-active {
	background-repeat: no-repeat;
    background-position-x: -100000vw;
}
```

ramonjd commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.css:
```diff
@@ -3,10 +3,47 @@
+	background-repeat: no-repeat;
+	background-position: center;
+	backdrop-filter: blur( 75px ) saturate( 50% );
+	-webkit-backdrop-filter: blur( 75px ) saturate( 50% );
```
The blur is causing some serious lag in desktop browsers. Do we need them?

apeatling replied:
I haven't had a chance to manage performance yet, so let me look at this today. Might need to change it out for a solid color or gradient.

glendaviesnz replied:
this approach was reverted

andrewserong commented:
Just noticed on mobile (while logged into my site) that it looks like the logged-in header covers up the close button? I'm wondering if we need to add back in one of the higher z-index values on the main container?

![image](https://user-images.githubusercontent.com/14988353/123055839-a45a1400-d449-11eb-91c5-5a79a6b5314a.png)

anomiex commented:
I find myself concerned that we're including a copy of the library in the repository, rather than pulling it in with the package manager (like we already do in the Slideshow block). We're generally trying to use package manager packages as it makes things more maintainable.

I see mention in p1624403130346300/1624401620.345200-slack-CBG1CP4EN that "we are doing a custom build of the swiper.js library to reduce the size by only including the modules we are using". How do we intend to keep this custom build up to date? For example, renovate informs us as package manager dependencies need updating, but I don't think it will see this custom build. And how does one actually create such a build when an upgrade is required? I see no instructions being included as part of the PR.

apeatling commented:
I'll work on fixing the remaining tasks for backgrounds in here today.

apeatling commented:
I've changed this from a background image blurred to a solid color that is the average of the image. The performance is much improved. 👍

ramonjd commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:
```diff
@@ -1624,15 +1259,53 @@
-				domUtil.emitEvent( carousel.container, 'jp_carousel.afterOpen' );
-			} );
+			// Need to set the overlay manually to block or swiper does't initialise properly.
+			carousel.overlay.style.opacity = 0;
```
Should this be `carousel.overlay.style.opacity = 1;` ? 

Otherwise the black background doesn't show.

glendaviesnz replied:
It is set to 1 by the fadeIn method further down.

ramonjd commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:
```diff
@@ -49,11 +38,46 @@
+			width = canvas.width = imgEl.naturalWidth || imgEl.offsetWidth || imgEl.width;
+
+			context.drawImage( imgEl, 0, 0 );
+			imgData = context.getImageData( 0, 0, width, height );
```
Do we need a `imgEl.crossOrigin = "Anonymous";` above this line?

I'm seeing `'CanvasRenderingContext2D': The canvas has been tainted by cross-origin data.` in my testing (desktop)

ramonjd commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:
```diff
@@ -49,11 +38,31 @@
+				return rgb;
+			}
+			canvas.width = 1;
+			canvas.height = 1;
```
Still think we might need `imgEl.crossOrigin = 'Anonymous';` ?

glendaviesnz commented:
> I find myself concerned that we're including a copy of the library in the repository

Valid points @anomiex, we will take a look to see what options there are for getting the benefits of the smaller custom build while also keeping a suitable level of automation to handle library updates.

glendaviesnz commented:
> I see no instructions being included as part of the PR.

@anomiex, @jeherve I don't think we are going to get to a fully automated state for the swiper.js custom build as part of this PR, but I have added documentation about [the process for manually generating the build](https://github.com/Automattic/jetpack/pull/20107/commits/32d79d4deacfeffa2397ef8320883b8b39bf6585), and have also added an issue which we will add to our board so we can follow up and explore ways to better automate this process.

jeherve commented:
I just noticed an issue when toggling the Comments option off in the Jetpack settings:

![image](https://user-images.githubusercontent.com/426388/123393083-0af04680-d59e-11eb-993a-25723a8ca139.png)

![image](https://user-images.githubusercontent.com/426388/123393264-3410d700-d59e-11eb-9232-5e5fac27711a.png)
`Uncaught TypeError: Cannot read property 'classList' of null`

stacimc commented:
Fix in now for the errors caused by toggling off either comments or info in the Jetpack settings 👍

samiff reviewed (commented):
I'm testing on a iPhone 12 mini (iOS 14.6) on [this](https://at.samiff.blog/2021/06/gallery-test-20107/) WoA site. Two things I've noticed:

1. When `Enable site accelerator` (Photon) is turned on, the image sizes in lightbox don't seem correct. Additionally, double-tapping to zoom on mobile Chrome zooms the whole page.
2. With Photon turned off, the Safari experience is very smooth and image sizes look great. On Chrome, I notice that the sliding action seemed choppy to to me. Also in Chrome, double tapping to zoom required me to tap _very_ quickly compared to Safari.

https://user-images.githubusercontent.com/15803018/123479468-38e58300-d5be-11eb-8354-5b9d45b8a915.mov

samiff commented:
General non-blocking design feedback: Would it be possible to have an indication that an image in the lightbox has comments? Otherwise you would have to click to find out for each image, as the comments section closes when navigating between images.

apeatling commented:
> When Enable site accelerator (Photon) is turned on, the image sizes in lightbox don't seem correct. Additionally, double-tapping to zoom on mobile Chrome zooms the whole page.

This should be fixed now (see https://github.com/Automattic/jetpack/pull/20204)

apeatling commented:
✅ Rebased

apeatling commented:
We're done with code merges from other PRs now. This is ready for freeze and review.

samiff commented:
@apeatling Just giving this another try now ([JN site](https://surprising-lizard.jurassic.ninja/2021/06/25/gallery-test-jn-20107/)) and noticed two items:

1. When opening a Gallery widget in a widget area I'm seeing `Uncaught TypeError: Cannot set property 'innerHTML' of null` and I'm unable to navigate with arrows or close the lightbox.
2. With Photon turned on, sizes seem to work well for `wp:gallery`, but images looks small on both `wp:jetpack/tiled-gallery` and `[gallery]` shortcode (video below).

https://user-images.githubusercontent.com/15803018/123700840-3ed99f00-d81e-11eb-8357-29751f1f57a0.mov

glendaviesnz commented:
> 1. When opening a Gallery widget in a widget area I'm seeing `Uncaught TypeError: Cannot set property 'innerHTML' of null`

@samiff this should be fixed by https://github.com/Automattic/jetpack/pull/20107/commits/c12ea77a97251e7f468e3f26d4ffb0cde42cfb33

sgomes commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:
```diff
@@ -1624,19 +1383,64 @@
 		// Register the event listener for starting the gallery
 		document.body.addEventListener( 'click', function ( e ) {
+			var isIE11 = window.navigator.userAgent.match( /Trident\/7\./ );
+			// IE11 support is being dropped in August 2021. The new swiper.js libray is not IE11 compat
```
@jeherve : Is this compatible with Jetpack's policies for a point release?

sgomes replied:
Nitpick: s/libray/library/

jeherve replied:
We're making a bit of an exception with IE11 right now, so that should be fine, yes.

sgomes commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:
```diff
@@ -1624,19 +1383,64 @@
 
 		// Register the event listener for starting the gallery
 		document.body.addEventListener( 'click', function ( e ) {
+			var isIE11 = window.navigator.userAgent.match( /Trident\/7\./ );
```
It's generally discouraged to use user agent parsing to identify individual browsers, instead of performing feature detection for the required features.

Since IE 11 doesn't support CSS grid, and since the new carousel requires it to work correctly, we can achieve a similar result by checking for feature support for grid, with the bonus that this will also catch other, older browsers, or any IE 11 instances masquerading as something else. Here's the code I'd use (note that I renamed the variable):

```suggestion
			var isCompatible = window.CSS && window.CSS.supports && window.CSS.supports( 'display', 'grid' );
```

apeatling replied:
Addressed in https://github.com/Automattic/jetpack/pull/20219

sgomes replied:
Thanks, @apeatling!

sgomes commented on projects/plugins/jetpack/modules/carousel/swiper-bundle.css:
```diff
@@ -0,0 +1,380 @@
+ */
+
+@font-face {
+  font-family: 'swiper-icons';
```
Does Swiper have an option for using plain SVGs or an SVG spritemap instead? As I mentioned above, icon fonts are a bad practice for a number of reasons.

glendaviesnz replied:
Update for this here https://github.com/Automattic/jetpack/pull/20222

sgomes replied:
Thank you, @glendaviesnz! 👍

sgomes reviewed (changes requested):
Thank you for all the work on the carousel! I'm looking forward to all of the usability improvements!

This is an initial, source-only review, to unblock anyone working on this. I'll be running some tests later today to see it in action, and will point out any issues I find separately.

jeherve commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.css:
```diff
@@ -344,7 +257,7 @@ div.jp-carousel-buttons a:hover {
 div.jp-carousel-buttons a.jp-carousel-commentlink,
 a.jp-carousel-image-download {
-	background: url(./images/carousel-sprite.png?5) no-repeat;
+	background: url( ./images/carousel-sprite.png?5 ) no-repeat;
```
That's a good point. It may be safer to build the path from a value passed via `wp_localize_script`.

glendaviesnz replied:
It currently works this way on WPcom (https://opengrok.a8c.com/source/xref/trunk/wp-content/mu-plugins/carousel/jetpack-carousel.css?r=227547#346) We haven't changed this part of the code at all, the css has just been formatted to add the spaces.  Checking on WPcom with current carousel it is working with this path with the wp-content path prepended:
<img width="604" alt="Screen Shot 2021-06-30 at 9 38 52 AM" src="https://user-images.githubusercontent.com/3629020/123871308-aeb76900-d987-11eb-8351-809f50ccc39f.png">

glendaviesnz replied:
It could still be worth exploring another approach to this at some point, eg. in the case of WPcom mapping to images on  CDN, so have added an issue for this https://github.com/Automattic/jetpack/issues/20221

sgomes replied:
You're absolutely right that the code was already there, @glendaviesnz, I don't know how I missed that! 🤦 Sorry about that!

Thank you for creating that issue 👍

jeherve reviewed (approved):
Nice work, this makes for a nice improvement over the existing Carousel modal! ❤️

I'd be tempted to merge this now, as it is, while making sure we address @sgomes's feedback in follow-up PRs. It will allow to get those changes in front of Beta testers sooner, and we can continue to iterate / fix bugs during the Beta week based on their feedback. It should also make it easier to review smaller PRs.

Approving this.

sgomes commented:
> I'd be tempted to merge this now, as it is, while making sure we address @sgomes's feedback in follow-up PRs.

Thank you, @jeherve! I'm tentatively okay with this (haven't had a chance to test things yet, sorry), as long as we make sure to hold off on merging to wpcom. Some of the issues that I pointed out need to be fixed before we can do that.

sgomes commented:
> Great news! One last step: head over to your WordPress.com diff, D63016-code, and commit it.
> Once you've done so, come back to this PR and add a comment with your changeset ID.
> 
> **Thank you!**

No, bad bot! 😄 

Jokes aside, please see the above comments. There are some pending issues that will need to be addressed before we can merge to wpcom, and will be addressed in future PRs.

apeatling commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:
```diff
@@ -391,45 +304,25 @@
-				carousel.nextButton = carousel.container.querySelector( '.jp-carousel-next-button' );
-				carousel.prevButton = carousel.container.querySelector( '.jp-carousel-previous-button' );
-				carousel.commentField = carousel.container.querySelector(
+				carousel.info = document.querySelector( '.jp-carousel-info' );
```
Addressed in #20219

sgomes replied:
Thank you!

apeatling commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.css:
```diff
@@ -977,14 +1081,19 @@ textarea#jp-carousel-comment-form-comment-field:focus::-webkit-input-placeholder
 
 	.jp-carousel-slide img {
-		filter: alpha(opacity=100);
+		filter: alpha( opacity=100 );
```
Addressed in #20219

sgomes replied:
👍

apeatling commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.css:
```diff
@@ -713,72 +736,79 @@ textarea#jp-carousel-comment-form-comment-field:focus::-webkit-input-placeholder
+#jp-carousel-comment-form-button-submit:active,
+#jp-carousel-comment-form-button-submit:focus {
+	background: white;
+    color: rgba( 34, 34, 34, 0.9 );
```
Addressed in #20219

apeatling replied:
Addressed in #20219

sgomes replied:
👍

apeatling commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.css:
```diff
@@ -653,58 +666,68 @@ textarea#jp-carousel-comment-form-comment-field:focus::-webkit-input-placeholder
 	-webkit-animation: load8 1.1s infinite linear;
 	animation: load8 1.1s infinite linear;
+	margin: 0 auto;
+    position: absolute; /* relative to .jp-carousel-comment-form-container */
```
Addressed in #20219

sgomes replied:
👍

apeatling commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.css:
```diff
@@ -911,55 +935,135 @@ textarea#jp-carousel-comment-form-comment-field:focus::-webkit-input-placeholder
-	border-top: 1px solid #f0f0f1;
+.jp-carousel-light .jp-carousel-fadeaway {
+	background: -moz-linear-gradient( bottom, rgba( 255, 255, 255, 0.75 ), rgba( 255, 255, 255, 0 ) );
+	background: -webkit-gradient(
```
Addressed in #20219

sgomes replied:
Thank you!

apeatling commented:
@sgomes I've addressed a number of the items in #20219

sgomes commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.css:
```diff
@@ -354,50 +267,108 @@ div.jp-carousel-buttons a.jp-carousel-commentlink {
 	div.jp-carousel-buttons a.jp-carousel-commentlink,
 	a.jp-carousel-image-download {
-		background-image: url(./images/carousel-sprite-2x.png?5);
+		background-image: url( ./images/carousel-sprite-2x.png?5 );
```
Resolved, see above.

sgomes commented:
Thank you for taking care of those fixes, @apeatling! 🙇
